File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/validator/common/update/AbstractTooCloseUpdateEpsilonValidator.java
package cz.fb.manaus.reactor.betting.validator.common.update;

import com.google.common.collect.Range;
import cz.fb.manaus.reactor.betting.BetContext;
import cz.fb.manaus.reactor.betting.validator.ValidationResult;
import cz.fb.manaus.reactor.betting.validator.Validator;

public abstract class AbstractTooCloseUpdateEpsilonValidator implements Validator {
    private final double epsilon;

    protected AbstractTooCloseUpdateEpsilonValidator(double epsilon) {
        this.epsilon = epsilon;
    }

    @Override
    public ValidationResult validate(BetContext context) {
        var oldOne = context.getOldBet().get().getRequestedPrice().getPrice();
        var newOne = context.getNewPrice().get().getPrice();
        var epsilon = (oldOne - 1) * this.epsilon;
        var closeRange = Range.closed(oldOne - epsilon, oldOne + epsilon);
        return ValidationResult.of(!closeRange.contains(newOne));
    }

    @Override
    public boolean isUpdateOnly() {
        return true;
    }

}

File: mns-reactor/src/main/java/cz/fb/manaus/reactor/betting/validator/common/update/AbstractTooCloseUpdateEpsilonValidator.kt
package cz.fb.manaus.reactor.betting.validator.common.update

import cz.fb.manaus.core.provider.ProviderTag
import cz.fb.manaus.reactor.betting.BetEvent
import cz.fb.manaus.reactor.betting.validator.ValidationResult
import cz.fb.manaus.reactor.betting.validator.Validator

abstract class AbstractTooCloseUpdateEpsilonValidator(private val epsilon: Double) : Validator<BetEvent> {

    override fun validate(event: BetEvent): ValidationResult {
        val oldOne = event.oldBet!!.requestedPrice.price
        val newOne = event.newPrice!!.price
        val epsilon = (oldOne - 1) * this.epsilon
        return if (newOne in oldOne - epsilon..oldOne + epsilon) ValidationResult.NOP else ValidationResult.ACCEPT
    }

    override val isUpdateOnly: Boolean = true

    override val tags get() = setOf(ProviderTag.PriceShiftContinuous)
}

